File: test/linux/werkplaats2/werkplaats2.py
import sys
import time
import pysoem

def main():
    if len(sys.argv) < 4:
        print(f'Usage: {sys.argv[0]} <network_interface> <slave_index> <channel_id>')
        return 1

    netif = sys.argv[1]
    slave_index = int(sys.argv[2])
    channel_index = int(sys.argv[3])

    if slave_index < 1:
        print(f'Invalid slave index: {slave_index}')
        return 1

    if channel_index < 1 or channel_index > 8:
        print(f'Invalid channel index: {channel_index}')
        return 1

    master = pysoem.Master()
    master.open(netif)

    try:
        if not master.config_init() or len(master.slaves) == 0:
            print(f'No slaves found or configuration failed on {netif}.')
            master.close()
            return 1

        print(f'{len(master.slaves)} slaves found and configured.')

        master.config_map()
        master.config_dc()

        print('Slaves mapped and configured for DC.')


        master.state = pysoem.OP_STATE
        master.write_state()
        
        while master.state_check(pysoem.OP_STATE, 50000) != pysoem.OP_STATE:
            print('Slaves not in operational state.')
            time.sleep(1)
            
        print('Slaves in operational state.')

        if master.state_check(pysoem.OP_STATE, 50000) == pysoem.OP_STATE:
            print('Operational state reached for all slaves.')
            while True:
                master.send_processdata()
                master.receive_processdata(10000)

                if slave_index > len(master.slaves):
                    print(f'Slave index out of range. Only {len(master.slaves)} slaves available.')
                    break

                slave = master.slaves[slave_index - 1]
                if slave.input:
                    # Extract the specified bit to check the sensor state
                    input_state = bool(slave.input[0] & (1 << (channel_index - 1)))
                    print(f'Input: {"true" if input_state else "false"}')

                time.sleep(1)  # Wait 1 second
        else:
            print(f'Failed to reach operational state. {master.state_check(pysoem.OP_STATE, 50000)}')

        master.close()
        return 0
    finally:
        master.state = pysoem.INIT_STATE
        # stuur alle slaves naar INIT state
        master.write_state()
        master.close()

if __name__ == '__main__':
    sys.exit(main())

File: test/linux/werkplaats2/werkplaats2.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include "ethercat.h"

// Declare IOmap as a global variable
uint8_t IOmap[4096];  // Adjust the size according to your needs

int main(int argc, char *argv[]) {
    if (argc < 4) {
        fprintf(stderr, "Usage: %s <network_interface> <slave_index> <channel_id>\n", argv[0]);
        return 1;
    }

    int slave_index = atoi(argv[2]);
    if (slave_index < 1) {
        fprintf(stderr, "Invalid slave index: %d\n", slave_index);
        return 1;
    }

    int channel_index = atoi(argv[3]);
    if (channel_index < 1 || channel_index > 8) {
        fprintf(stderr, "Invalid channel index: %d\n", channel_index);
        return 1;
    }

    if (!ec_init(argv[1])) {
        fprintf(stderr, "Error initializing the EtherCAT master on %s\n", argv[1]);
        return 1;
    }

    printf("EtherCAT initialization on %s succeeded.\n", argv[1]);

    if (ec_config_init(FALSE) <= 0) {
        fprintf(stderr, "No slaves found or configuration failed.\n");
        ec_close();
        return 1;
    }

    printf("%d slaves found and configured.\n", ec_slavecount);

    ec_config_map(&IOmap);
    ec_configdc();

    printf("Slaves mapped and configured for DC.\n");

    ec_statecheck(0, EC_STATE_SAFE_OP, EC_TIMEOUTSTATE * 4);

    ec_slave[0].state = EC_STATE_OPERATIONAL;
    ec_writestate(0);

    while (ec_statecheck(0, EC_STATE_OPERATIONAL, EC_TIMEOUTSTATE) != EC_STATE_OPERATIONAL) {
        fprintf(stderr, "Slaves not in operational state.\n");
        sleep(1);
    }

    printf("Operational state reached for all slaves.\n");

    while (1) {
        ec_send_processdata();
        ec_receive_processdata(EC_TIMEOUTRET);

        if (slave_index > ec_slavecount) {
            fprintf(stderr, "Slave index out of range. Only %d slaves available.\n", ec_slavecount);
            break;
        }

        // Assuming EL1008 starts at position 0 in the input bytes
        uint8_t inputs = *(ec_slave[slave_index].inputs);
        
        printf("Input: %s\n", (inputs & (1 << (channel_index - 1))) ? "true" : "false");

        sleep(1);  // Wait 1 second
    }

    ec_close();
    return 0;
}
